airbyte-integrations/connectors/source-posthog/source_posthog/streams.py

```diff
@@ -0,0 +1,306 @@
+                    self.reversed_pagination["is_completed"] = True
+                    yield from []
+                elif state_value < last_record_value:  # full page
+                    self.reversed_pagination["latest_response_state"] = {self.cursor_field: last_record_value}
+                    yield from data
+                else:  # last_record_value <= state_value < first_record_value
+                    slice_ = [i for i in data if i[self.cursor_field] > state_value]
+                    self.reversed_pagination["latest_response_state"] = {self.cursor_field: slice_[-1][self.cursor_field]}
+                    self.reversed_pagination["is_completed"] = True
+                    yield from slice_
+
+        else:
+            yield from data
+
+
+class Annotations(IncrementalPosthogStream):
+    """
+    OK sorting by updated_at exists, but we still does not have `since`-like params for this endpoint.
+    Hence sorting exists it is possible to make it incremental, but this will be usual reverse incremental
+    from latest to state value.
+    """
+
+    cursor_field = "updated_at"
+
+    def path(self, **kwargs) -> str:
+        return "annotation"
+
+    def request_params(self, stream_state: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+
+        stream_state = stream_state or {}
+        params = super().request_params(stream_state=stream_state, **kwargs)
+        params["order"] = f"-{self.cursor_field}"
+        return params
+
+
+class Cohorts(IncrementalPosthogStream):
+    """
+    normal ASC sorting. But without filters like `since`
+    So we need to query all anyway, but we may skip already written reords for incremental.
+    """
+
+    cursor_field = "id"
+
+    def path(self, **kwargs) -> str:
+        return "cohort"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        resp_json = response.json()
+        if "next" in resp_json and resp_json["next"]:
+            next_query_string = urllib.parse.urlsplit(resp_json["next"]).query
+            params = dict(urllib.parse.parse_qsl(next_query_string))
+            if params:
+                return params
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+        current_stream_state = current_stream_state or {}
+        latest_state = latest_record.get(self.cursor_field)
+        current_state = current_stream_state.get(self.cursor_field) or latest_state
+        return {self.cursor_field: max(latest_state, current_state)}
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
```

Comment:
I would like the hefty code in the `parse_response` method for the `IncrementalPosthogStream` class to take into account this logic as well, otherwise this code is a stripped-down version of the parent class method.

Reply:
currently such cases are covered:
1) full-refresh parse_response = just read from data_field
2) incremental parse_response = read with setting/modifying flags, DESC paginated order)
3) exception: cohorts - (ASC paginated order without flags - differs from DESC by state comparison) - does not need flags 
4) elements `/element/stats` - exception out of the usual rules